Create docker-image.yml

dragonruler1000 opened the pull request:
Added a workflow to build the docker image automatically.

dragonruler1000 committed:
Create docker-image.yml

Added a workflow to build the docker image automatically.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Amulet-Team/Amulet-Map-Editor?pullRequest=1184) <br/>All committers have signed the CLA.

dragonruler1000 committed:
Add files via upload

gentlegiantJGC commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,29 @@
+on:
+  push:
+    tags:
+      - "v*"  # Trigger on version tags like v1.0, v0.10, etc.
```
We don't add a `v` prefix to version numbers.
See here for how to implement this
https://github.com/Amulet-Team/Amulet-Map-Editor/blob/4889915b0f7a30416b9994929241dc657f6848b4/.github/workflows/python-build.yml#L6C1-L8C23

dragonruler1000 replied:
updated on my branch.

gentlegiantJGC commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,29 @@
+      with:
+        files: amulet-editor.tar
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
This can be implemented with this github command
https://github.com/Amulet-Team/Amulet-Map-Editor/blob/4889915b0f7a30416b9994929241dc657f6848b4/.github/workflows/python-build.yml#L66C1-L69C108
Docs
https://cli.github.com/manual/gh_release_upload

gentlegiantJGC commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+RUN apt-get update && apt install libgtk-3-dev dbus-x11 wget libnotify4 -y
+RUN pip install --upgrade pip
+RUN wget https://extras.wxpython.org/wxPython4/extras/linux/gtk3/debian-11/wxPython-4.2.0-cp39-cp39-linux_x86_64.whl && pip install wxPython-4.2.0-cp39-cp39-linux_x86_64.whl
+RUN pip install amulet-map-editor==0.10.44 
```
I would prefer if this was not hard coded.
You have already cloned the repository so can you do `pip install .` to install from the current directory?
If docker changes the working directory you may need to change the path.

dragonruler1000 replied:
I didn't realise that it was hard coded. I think that was just because i had copied the info from pypi.  updated it in the docker file on my fork so it should update in the PR.

gentlegiantJGC commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM python:3.9-bullseye
+RUN apt-get update && apt install libgtk-3-dev dbus-x11 wget libnotify4 -y
+RUN pip install --upgrade pip
+RUN wget https://extras.wxpython.org/wxPython4/extras/linux/gtk3/debian-11/wxPython-4.2.0-cp39-cp39-linux_x86_64.whl && pip install wxPython-4.2.0-cp39-cp39-linux_x86_64.whl
```
Where are you requesting the version of Python?
Is 3.9 the default for the runner?
You can use `actions/setup-python@v5` to ensure a specific version of python is installed
https://github.com/Amulet-Team/Amulet-Map-Editor/blob/4889915b0f7a30416b9994929241dc657f6848b4/.github/workflows/python-build.yml#L25C1-L29C53

dragonruler1000 replied:
The verison of python is set in the FROM line so yes python 3.9 is the version of python that it is using.

dragonruler1000 committed:
Update Dockerfile

dragonruler1000 committed:
Update docker-image.yml

dragonruler1000 committed:
Update docker-image.yml

dragonruler1000 reviewed (commented):
Fixed/implemented all of your issues and suggestions.

gentlegiantJGC commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,28 @@
+      env:
+        GITHUB_TOKEN: ${{ github.TOKEN }}
+      run: |
+        gh release upload ${{github.event.release.tag_name}} ./dist/${{env.BUILT_PATH}}#${{env.BUILT_PATH}}
```
`./dist/${{env.BUILT_PATH}}#${{env.BUILT_PATH}}`
This path is incorrect.
The part before the `#` is the path to upload and the part after is the file name the user will see when downloading.
`env.BUILT_PATH` was a variable set in the workflow you copied it from.

I suggest something like this `amulet-editor.tar#Amulet-v${{ github.event.release.tag_name }}-os-architecture.tar`
Swap out `os-architecture` for sensible values so the user knows which version to download.

gentlegiantJGC commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+RUN apt-get update && apt install libgtk-3-dev dbus-x11 wget libnotify4 -y
+RUN pip install --upgrade pip
+RUN wget https://extras.wxpython.org/wxPython4/extras/linux/gtk3/debian-11/wxPython-4.2.0-cp39-cp39-linux_x86_64.whl && pip install wxPython-4.2.0-cp39-cp39-linux_x86_64.whl
+RUN pip install amulet-map-editor 
```
This does not solve the original issue. The new version may not have been uploaded to pypi when this command is run.
`pip install .` will install it from the current directory.

dragonruler1000 replied:
Fixed this by having the workflow run after python-build is completed so it is will have been pushed to pypi.

gentlegiantJGC replied:
why not use the command I sent?

dragonruler1000 replied:
because im also wanting the docker file to be able to be built by the user after they download it. because when you run the docker image you are running the Dockerfile and if people want to update the docker image they can just run docker build path/to/Dockerfile and i don't feel like having the Dockerfile clone the git hub repo is going to be a good way to have it build it.

gentlegiantJGC replied:
Okay. I have no experience with docker

dragonruler1000 committed:
Update docker-image.yml

dragonruler1000 committed:
Update docker-image.yml

step 1

dragonruler1000 committed:
Update docker-image.yml

dragonruler1000 committed:
Update docker-image.yml

dragonruler1000 reviewed (commented):
That should fix all the issues.

dragonruler1000 committed:
Update Dockerfile

making it so it will update the install version of amulet when built. before it wouldn't do anything unless the Docker file was manually edited.

dragonruler1000 committed:
Update docker-image.yml

did the other side of the code

dragonruler1000 committed:
Update Dockerfile

changing fall back version

dragonruler1000 commented:
Just did some commits that will make it so Docker will build the latest version of amulet and not use the cached version every time.

gentlegiantJGC commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+RUN pip install --upgrade pip
+RUN wget https://extras.wxpython.org/wxPython4/extras/linux/gtk3/debian-11/wxPython-4.2.0-cp39-cp39-linux_x86_64.whl && pip install wxPython-4.2.0-cp39-cp39-linux_x86_64.whl
+ARG AMULET_VERSION=0.10.44  # defaults to version 0.10.44 if no arg is given in build command.
+RUN pip install amulet-map-editor==$AMULET_VERSION
```
I like the variable configuration but I don't like the hard coded default value. I would prefer to set this and forget about it. I would also like to support alpha and beta releases.

Does docker have an if else statement?
Something like
```
ARG AMULET_VERSION=RELEASE
IF AMULET_VERSION==RELEASE {
    RUN pip install --upgrade --upgrade-strategy eager amulet-map-editor
} ELIF AMULET_VERSION==BETA {
    RUN pip install --upgrade --upgrade-strategy eager amulet-map-editor>=0b0
} ELIF AMULET_VERSION==ALPHA {
    RUN pip install --upgrade --upgrade-strategy eager amulet-map-editor>=0a0
} ELSE {
    RUN pip install amulet-map-editor==$AMULET_VERSION
}
```

I think that should work

dragonruler1000 replied:
I'll probably just leave it blank. Yes, there is a way to have an if-else statement but it is more complicated than I want to try. I don't think most people are going to care if they can just install the beta version by passing BETA into the argument. If someone wants to do that they can add the functionality themselves.

dragonruler1000 replied:
Or if passing obo/oba doesn't work via the command line you can always just manually edit the Docker file to get it to work.

dragonruler1000 committed:
Update Dockerfile

dragonruler1000 committed:
Update Dockerfile

gentlegiantJGC reviewed (approved):
Looks good. I will do an alpha release to test it.

gentlegiantJGC commented:
Here is the release. If everything goes well the file should appear here
https://github.com/Amulet-Team/Amulet-Map-Editor/releases/tag/0.10.45a0

This is the main build
https://github.com/Amulet-Team/Amulet-Map-Editor/actions/runs/16726368107

Your new workflow should appear here when that completes
https://github.com/Amulet-Team/Amulet-Map-Editor/actions/workflows/docker-image.yml

gentlegiantJGC commented:
@dragonruler1000 
```
#9 [5/6] RUN pip install amulet-map-editor==
#9 0.389 ERROR: Invalid requirement: 'amulet-map-editor==': Expected end or semicolon (after name and no valid version specifier)
#9 0.389     amulet-map-editor==
#9 0.389                      ^
#9 ERROR: process "/bin/sh -c pip install amulet-map-editor==$AMULET_VERSION" did not complete successfully: exit code: 1
------
 > [5/6] RUN pip install amulet-map-editor==:
0.389 ERROR: Invalid requirement: 'amulet-map-editor==': Expected end or semicolon (after name and no valid version specifier)
0.389     amulet-map-editor==
0.389                      ^
------
Dockerfile:6
--------------------
   4 |     RUN wget https://extras.wxpython.org/wxPython4/extras/linux/gtk3/debian-11/wxPython-4.2.0-cp39-cp39-linux_x86_64.whl && pip install wxPython-4.2.0-cp39-cp39-linux_x86_64.whl
   5 |     ARG AMULET_VERSION  # defaults to the latest version if no arg is given in build command.
   6 | >>> RUN pip install amulet-map-editor==$AMULET_VERSION
   7 |     RUN apt-get install libSDL2-2.0 -y
   8 |     
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c pip install amulet-map-editor==$AMULET_VERSION" did not complete successfully: exit code: 1
```

gentlegiantJGC commented:
I think the issue is that the workflow does not have access to `github.event.release.tag_name` because it isn't run from a publish event.

gentlegiantJGC commented:
Here is the full output.
[job-logs.txt](https://github.com/user-attachments/files/21580528/job-logs.txt)

dragonruler1000 commented:
I know im trying to fix it right now. I'm going to open a new pull request in a little bit that will hopefully fix the problem.

dragonruler1000 commented:
@gentlegiantJGC should be fixed by #1187
